keps/sig-testing/5468-invariant-testing/README.md

```diff
@@ -0,0 +1,960 @@
+to flaking invariant tests in a timeline fashion will result in demoting or
+removing them.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+If implemented poorly, this could result in tests flaking in any number of e2e
+test CI jobs that are now running these tests.
+
+We will mitigate this by thoroughly reviewing and testing the invariant checking utilities, 
+and by limiting to only portable, reliable mechanisms.
+
+If not run in many CI jobs, there will be limited benefit to the signal.
+We will aim to generally introduce these as default selected tests.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+[ginkgo], which we use for the e2e tests, does not have a facility for 
+"this test must run after all other tests". It does have ordering, but only
+within a group of tests. Alternatively we can run arbitrary code after the
+suite, but that does not suit our ability to opt-in/out of the tests.
+
+We will introduce a package under `test/e2e` that contains sentinel test(s) which
+do nothing other than register to ginkgo that they have run, if selected.
+
+These tests can be labeled appropriately, allowing the existing test selection
+mechanisms to select / skip them.
+
+We can then use `ginkgo.ReportAfterSuite` to inspect which, if any, of these 
+sentinel invariant tests ran, and run the *actual* test logic, reporting pass/fail.
+
+This allows us the flexibility of treating them like regular tests while ensuring
+that the actual invariant check occurs _after_ all other tests.
+
+It also ensures that while default-enabled, these tests will not impact
+conformance testing, since the sentinel tests will not be tagged `[Conformance]`
+and therefore will not run, in turn disabling the actual after-suite checks.
+
+Invariant checks will be limited to read-only, such as fetching metrics.
+Invariant checks should be simple, reliable, cheap code.
+They ideally run in most e2e test runs for useful coverage.
+
+Failure logs for these invariant checks will include instructions about how
+to handle the test failure and who to report bugs to, from the documented owners.
+
+Something like this, for the failure message:
+> Invariant failed for missing metric: <metric-name>
+> If this failed on a PR, check your PR for bugs, if this failed on a periodic test please file a bug and assign the owners.
```

Comment:
Isn't "check your PR for bugs" the wrong message and in conflict with "tragedy of the commons"? It implies that it is the responsibility of the PR owner instead of the invariant owner to figure out why the invariant was violated. If we go down this route, it becomes the responsibility of everyone to debug the feature tested by the invariants.

Invariants are most likely to break because of some change, i.e. in a PR, unless we merge some changes where tests only run in CI or in additional environments.

Reply:
OTOH, the invariant might really have been broken by the PR.

How about this?

```
If this failed, search for an existing GitHub issue. It might be a known problem
that you can ignore.

Otherwise, if this failed on a periodic job, please file a bug and assign the owners.
If this failed on a PR, try to determine if it is something that could have been
caused by your changes. If not or unsure, file an issue.
```

Reply:
I agree that some due diligence by the PR author is needed.

I'm not sure whether you saw my comment with the "OTOH" (https://github.com/kubernetes/enhancements/pull/5514#discussion_r2333783962). I was suggesting an update of the wording there.

Reply:
I think I have already covered the requirements for that without the sample message.